narwhal/executor/src/subscriber.rs

```diff
@@ -243,151 +275,311 @@ impl<Network: SubscriberNetwork> Fetcher<Network> {
-            stagger += Duration::from_millis(200);
+            // TODO: Can further parallelize this by worker_id if necessary.
+            // Only have one worker for now so will leave this for a future
+            // optimization.
+            let mut remaining_digests = digests.clone();
+
+            let local_batches = self
+                .try_fetch_locally(remaining_digests.clone().into_iter().collect(), worker_id)
+                .await;
+            for (local_batch_digest, local_batch) in local_batches {
+                if remaining_digests.remove(&local_batch_digest) {
+                    let batch_fetch_duration =
+                        local_batch.metadata.created_at.elapsed().as_secs_f64();
+                    self.metrics
+                        .batch_execution_latency
+                        .observe(batch_fetch_duration);
+                    debug!(
+                        "Batch {local_batch_digest:?} took {batch_fetch_duration} seconds to be fetched for execution since creation",
+                    );
+                    fetched_batches.insert(local_batch_digest, local_batch);
+                }
+            }
+
+            if remaining_digests.is_empty() {
+                return fetched_batches;
+            }
+
+            let _timer = self.metrics.subscriber_remote_fetch_latency.start_timer();
+            let mut stagger = Duration::from_secs(0);
+            let mut futures = FuturesUnordered::new();
+            for worker in &workers {
+                let future = self.fetch_remote(stagger, worker.clone(), remaining_digests.clone());
+                futures.push(future.boxed());
+                // TODO: Make this a parameter, and also record workers / authorities that are down
+                // to request from them batches later.
+                stagger += Duration::from_millis(500);
+            }
+
+            while let Some(remote_batches) = futures.next().await {
+                for (remote_batch_digest, remote_batch) in remote_batches {
+                    if remaining_digests.remove(&remote_batch_digest) {
+                        let batch_fetch_duration =
+                            remote_batch.metadata.created_at.elapsed().as_secs_f64();
+                        self.metrics
+                            .batch_execution_latency
+                            .observe(batch_fetch_duration);
+                        debug!(
+                            "Batch {remote_batch_digest:?} took {batch_fetch_duration} seconds to be fetched for execution since creation"
+                        );
+
+                        fetched_batches.insert(remote_batch_digest, remote_batch);
+                    }
+                }
+
+                if remaining_digests.is_empty() {
+                    return fetched_batches;
+                }
+            }
+
+            // Should never reach here, as we should keep looping until we have all batches remotely or locally.
```

Comment:
I think we will hit this log if there are more than 1 worker? I can't find the retry loop after exhausting all remote workers to fetch from.